File: python/duplicate_zeros.py
from typing import List


class Solution:
    def duplicateZeros(self, array: List[int]) -> None:
        removals, k, i = 0, len(array) - 1, 0
        while i < len(array) - removals:
            if array[i] is 0:
                if i == len(array) - removals - 1:
                    array[k] = 0
                    k -= 1
                removals += 1
            i += 1

        i = len(array) - removals - 1
        while i >= 0:
            if array[i] is 0:
                array[k], array[k - 1] = 0, 0
                k -= 1
            else:
                array[k] = array[i]
            k -= 1
            i -= 1

File: java/src/DuplicateZeros.java
public class DuplicateZeros {
    public static void duplicateZeros(int[] array) {
        int removals = 0, k = array.length - 1;
        for (int i = 0 ; i < array.length - removals ; i++) {
            if (array[i] == 0) {
                if (i == array.length - removals - 1) {
                    array[k--] = 0;
                }
                removals++;
            }
        }
        for (int index = array.length - removals - 1 ; index >= 0 ; index--, k--) {
            if (array[index] == 0) {
                array[k] = array[k - 1] = 0;
                k--;
            } else {
                array[k] = array[index];
            }
        }
    }
}
